### PR TITLE
Adding Valgrind build task

### PR DESCRIPTION
**Pull request by sffc**
I wrote a build task that runs the examples through Valgrind Callgrind and saves the results.  This can be used as another source of instruction count measurements.  It obviously only works on platforms where Valgrind is available.  I added it to Makefile.toml so it is available for anyone who wants it, but I haven't yet added it to the CI.

### CODE REVIEW

**Comment by sffc** [comment]
Here are some example visualizations of the callgrind out files ([instructions](https://stackoverflow.com/a/14938582/1407170)):

code_line_diff
language_names_hash_map

File: Cargo.toml [hunk 1]
```diff
@@ -43,3 +43,4 @@ lto = true
 # https://github.com/rust-lang/cargo/issues/6988
 [profile.bench]
 debug = true
+debug-assertions = false
```
Comment (sffc) [inline Cargo.toml:46]: Note: I made this change because when running Valgrind, debug-assertions was producing a lot of noise.  This change also disables debug-assertions for the memory benchmarks, which is probably a good thing.

File: Makefile.toml [hunk 2]
```diff
@@ -286,6 +286,52 @@ category = "ICU4X WASM"
 command = "rm"
 args = ["-rf", "wasmpkg"]
 
+### VALGRIND TASKS ###
+
+[tasks.valgrind]
+description = "Build ICU4X example files with default features and run through Valgrind"
```
Comment (gregtatum) [inline Makefile.toml:292]: Suggestion: I think the valgrind command could use some documentation that explains what it is, how it works, and how to view the results. I believe `tools/benchmark/README.md` would be appropriate place to do this.

File: Makefile.toml [hunk 3]
```diff
@@ -286,6 +286,52 @@ category = "ICU4X WASM"
...
+    trigger_error "Build failed! To debug, build examples with `--features icu_benchmark_macros/rust_global_allocator`"
+end
+trimmed_stdout = trim ${output.stdout}
+json_messages = split ${trimmed_stdout} "\n"
+for json_message in ${json_messages}
+    json_obj = json_parse ${json_message}
+
+    is_compiler_artifact = eq ${json_obj.reason} "compiler-artifact"
+    is_example = eq ${json_obj.target.kind[0]} "example"
+    if ${is_compiler_artifact} and ${is_example}
```
Comment (gregtatum) [inline Makefile.toml:315]: Suggestion (optional): I would appreciate a few comments explaining what's going on here in general.

File: components/decimal/examples/code_line_diff.rs [hunk 4]
```diff
@@ -41,6 +41,8 @@ fn main(_argc: isize, _argv: *const *const u8) -> isize {
         let decimals: (FixedDecimal, FixedDecimal) = (line.0.into(), line.1.into());
         let removed = fdf.format(&decimals.0);
         let added = fdf.format(&decimals.1);
+        assert_ne!("", removed.writeable_to_string());
```
Comment (gregtatum) [inline components/decimal/examples/code_line_diff.rs:44]: Question: Are these changes supposed to be here? They look unrelated.

File: tools/benchmark/macros/Cargo.toml [hunk 5]
```diff
@@ -11,5 +11,10 @@ version = "0.1.0"
 [dependencies]
 dhat = { version = "0.2", optional = true }
 
+# This cfg originates in dlmalloc/lib.rs
+[target.'cfg(any(target_os = "linux", target_os = "macos", target_arch = "wasm32"))'.dependencies]
+dlmalloc = { version = "0.2", optional = true, features = ["global"] }
+
 [features]
 benchmark_memory = ["dhat"]
+rust_global_allocator = ["dlmalloc"]
```
Comment (gregtatum) [inline tools/benchmark/macros/Cargo.toml:20]: Question: Why is this called `rust_global_allocator`? Could it be possible that we want other global allocator configurations? Would it be better to call this something like `valgrind_allocator`?

Comment (gregtatum) [inline tools/benchmark/macros/Cargo.toml:20]: Suggestion: This is a new feature. We have #556 on file to document the features, but it would be good to either add it to that issue to follow-up, or stub out the feature documentation with this feature.

**Review by gregtatum — CHANGES_REQUESTED** [review]
Thanks, this looks like a pretty good approach to me. Most of my feedback was around documentation of the feature, rather than the approach.

File: Makefile.toml (hunk 2 above)
Comment (dminor) [reply Makefile.toml:292]: In case it's helpful, this is what I did locally:
```
pip install gprof2dot
gprof2dot --format=callgrind --output=out.dot benchmarks/valgrind/tui.out
dot -Tpng out.dot -o graph.png
eog graph.png
```

**Review by dminor — CHANGES_REQUESTED** [review]
Nice! This looks like it will be very helpful. I just have one issue around suppressing stderr that caused some difficulties for me.

File: Makefile.toml [hunk 6]
```diff
@@ -286,6 +286,52 @@ category = "ICU4X WASM"
...
+script = '''
+exit_on_error true
+
+valgrind = which valgrind
+assert ${valgrind} "Could not find 'valgrind' in path.\n*** Please run 'apt-get install valgrind' ***"
+
+mkdir benchmarks
+mkdir benchmarks/valgrind
+
+output = exec cargo +nightly-2021-02-28 build --examples --message-format=json --features icu_benchmark_macros/rust_global_allocator --profile bench -Z unstable-options
```
[the remark this answers is not in the document]
Comment (sffc) [reply Makefile.toml:304]: I used `nightly-2021-02-28` here because `nightly-2021-03-15` (as well as nightly head) are broken with WebAssembly.

[the remark this answers is not in the document]
Comment (sffc) [reply Makefile.toml:304]: Follow-up: #651

File: tools/benchmark/macros/src/lib.rs [hunk 7]
```diff
@@ -2,41 +2,63 @@
...
-        // The dhat instance will be alive for the life of the main function, and when dropped,
-        // it will output heap usage information.
-        let _dhat = Dhat::start_heap_profiling();
-        eprintln!("Feature");
-    };
+// Enable Dlmalloc if the "rust_global_allocator" feature is present alone
+#[cfg(all(feature = "rust_global_allocator", not(feature = "benchmark_memory")))]
+pub mod rust_global_allocator {
+    /// Re-export tools so that macros can access it without needing a dependency.
+    pub use dlmalloc;
```
[the remark this answers is not in the document]
Comment (sffc) [reply tools/benchmark/macros/src/lib.rs:50]: Added documentation to the README.md file.

File: Makefile.toml [hunk 8]
```diff
@@ -286,6 +286,52 @@ category = "ICU4X WASM"
...
+
+[tasks.valgrind]
+description = "Build ICU4X example files with default features and run through Valgrind"
+category = "ICU4X Valgrind"
+script_runner = "@duckscript"
+script = '''
+exit_on_error true
+
+valgrind = which valgrind
+assert ${valgrind} "Could not find 'valgrind' in path.\n*** Please run 'apt-get install valgrind' ***"
```
[the remark this answers is not in the document]
Comment (sffc) [reply Makefile.toml:299]: The Valgrind package claims to run on macOS and Solaris as well.  I haven't tried, but `brew install valgrind` might work:

https://formulae.brew.sh/formula/valgrind

File: Makefile.toml (hunk 2 above)
Comment (sffc) [reply Makefile.toml:292]: Added to README.md as suggested.

File: Makefile.toml (hunk 3 above)
Comment (sffc) [reply Makefile.toml:315]: Added comments.

File: components/decimal/examples/code_line_diff.rs (hunk 4 above)
Comment (sffc) [reply components/decimal/examples/code_line_diff.rs:44]: They're not *completely* unrelated.  I made these changes to improve the quality of the build artifacts in release mode.  It turns out that `assert_ne!()` with all the build settings does a good job at keeping things from being deleted by dead code elimination.  I could have done this in a separate PR, but I did it here because it was simpler.

File: tools/benchmark/macros/Cargo.toml (hunk 5 above)
Comment (sffc) [reply tools/benchmark/macros/Cargo.toml:22]: What @dminor said.  The point is to remove the system dependency.  There's no particular reason that this is `dlmalloc` (it could be any other Rust-based allocator), and it could be useful outside of Valgrind.

Comment (sffc) [reply tools/benchmark/macros/Cargo.toml:22]: Done.

File: Makefile.toml [hunk 9]
```diff
@@ -286,6 +286,52 @@ category = "ICU4X WASM"
...
+
+mkdir benchmarks
+mkdir benchmarks/valgrind
+
+output = exec cargo +nightly-2021-02-28 build --examples --message-format=json --features icu_benchmark_macros/rust_global_allocator --profile bench -Z unstable-options
+if ${output.code}
+    trigger_error "Build failed! To debug, build examples with `--features icu_benchmark_macros/rust_global_allocator`"
```
[the remark this answers is not in the document]
Comment (sffc) [reply Makefile.toml:306]: I fixed this by splitting the build task into a separate Makefile task, such that you can see all the errors in real time.  This is better than "cargo make valgrind" hanging for a long time while the build finishes.

**Review by dminor — APPROVED** [review]
lgtm!

File: tools/benchmark/macros/Cargo.toml (hunk 5 above)
Comment (gregtatum) [reply tools/benchmark/macros/Cargo.toml:24]: Ah, I _think_ the dhat benchmark sets its own custom global allocator too.

Comment (sffc) [reply tools/benchmark/macros/Cargo.toml:24]: Indeed it does, which is why these two features are mutually exclusive.

I tried just using the dhat global allocator, but it had too much overhead for it to be useful in performance benchmarks.

**Review by gregtatum — APPROVED** [review]
Thanks for the docs! This looks good to me.
